cpp/src/arrow/filesystem/localfs_benchmark.cc

```diff
@@ -0,0 +1,147 @@
+#include <arrow/testing/random.h>
+#include <arrow/util/async_generator.h>
+#include <arrow/util/string_view.h>
+#include "arrow/filesystem/localfs.h"
+#include "arrow/io/file.h"
+#include "arrow/table.h"
+#include "arrow/testing/future_util.h"
+#include "arrow/testing/gtest_util.h"
+#include "arrow/testing/random.h"
+#include "arrow/util/formatting.h"
+#include "arrow/util/io_util.h"
+#include "arrow/util/make_unique.h"
+#include "arrow/util/string_view.h"
+
+namespace arrow {
+
+namespace fs {
+
+using arrow::internal::make_unique;
+using arrow::internal::TemporaryDir;
+
+/// Set up hierarchical directory structure to test asynchronous
+/// file discovery interface (GetFileInfoGenerator()) in the LocalFileSystem
+/// class.
+///
+/// The main routine of the class is `InitializeDatasetStructure()`, which
+/// does the following:
+/// 1. Create `num_files_` empty files under specified root directory.
+/// 2. Create `num_dirs_` additional sub-directories in the current dir.
+/// 3. Check if the specified recursion limit is reached (controlled by `nesting_depth_`).
+///   a. Return if recursion limit reached.
+///   b. Recurse into each sub-directory and perform steps above, increasing current
+///      nesting level.
+class LocalFSFixture : public benchmark::Fixture {
+ public:
+  void SetUp(const benchmark::State& state) override {
+    ASSERT_OK_AND_ASSIGN(tmp_dir_, TemporaryDir::Make("localfs-test-"));
+
+    auto options = LocalFileSystemOptions::Defaults();
+    fs_ = make_unique<LocalFileSystem>(options);
+
+    InitializeDatasetStructure(0, tmp_dir_->path());
+  }
+
+  void InitializeDatasetStructure(size_t cur_nesting_level,
+                                  arrow::internal::PlatformFilename cur_root_dir) {
+    ASSERT_OK(arrow::internal::CreateDir(cur_root_dir));
+
+    arrow::internal::StringFormatter<Int32Type> format;
+    for (size_t i = 0; i < num_files_; ++i) {
+      std::string fname = "file_";
+      format(i, [&fname](util::string_view formatted) {
+        fname.append(formatted.data(), formatted.size());
+      });
+      ASSERT_OK_AND_ASSIGN(auto path, cur_root_dir.Join(std::move(fname)));
+      ASSERT_OK(MakeEmptyFile(path.ToString()));
+    }
+
+    if (cur_nesting_level == nesting_depth_) {
+      return;
+    }
+
+    for (size_t i = 0; i < num_dirs_; ++i) {
+      std::string dirname = "dir_";
+      format(i, [&dirname](util::string_view formatted) {
+        dirname.append(formatted.data(), formatted.size());
+      });
+      ASSERT_OK_AND_ASSIGN(auto path, cur_root_dir.Join(std::move(dirname)));
+      InitializeDatasetStructure(cur_nesting_level + 1, std::move(path));
+    }
+  }
+
+  Status MakeEmptyFile(const std::string& path) {
+    return io::FileOutputStream::Open(path).status();
+  }
+
+ protected:
+  std::unique_ptr<TemporaryDir> tmp_dir_;
+  std::unique_ptr<LocalFileSystem> fs_;
+
+  const size_t nesting_depth_ = 2;
+  const size_t num_dirs_ = 10;
+  const size_t num_files_ = 10000;
```

Comment:
This creates on the order of 1 million files total, and makes the benchmark rather slow to setup and run.

Reply:
The reason to create so many files was to adequately test for large batch/dir_readahead combinations, which I think show sensible results only if there's enough files to stress the target code. 

But, if you think we can reduce the overall number of files, I don't have any strong objections to that.

Reply:
`num_files_ = 1000` looks more reasonable here, but YMMV :-)

Reply:
Yeah, there's a bit of a tension between producing interesting numbers and making the benchmark rather expensive to run (even on a fast machine... I'm afraid macOS or Windows might be dismal here :-)).

Reply:
I agree, win and macOS users will almost surely suffer in this case...
So, to keep benchmarking fast, let `num_files_` be 1k, then. :)